uber/site_sections/registration.py

```diff
@@ -828,6 +828,70 @@ def discount(self, session, message='', **params):
 
         return {'message': message}
 
+    @csv_file
+    def generate_promo_codes(self, out, session, **params):
+        date = params['expiration_date'].split("-")
+        params['expiration_date'] = datetime(int(date[0]), month=int(date[1]), day=int(date[2]))
+        out.writerow(['Code', 'Price', 'Uses', 'Expiration Date'])
+        count = int(params['count'])
+        codes = []
+        for x in range(count):
+            code = session.promo_code(params)
+            session.add(code)
+            codes.append(code)
+        session.commit()
+        for code in codes:
+            uses = code.uses
+            if uses is None:
+                uses = "Unlimited"
+            out.writerow([code.code, code.price, uses, code.expiration_date])
+
+    @site_mappable
+    def generate_promo_code(self, session, message='', **params):
+        if 'code' in params:
+            date = params['expiration_date'].split("-")
+            params['expiration_date'] = datetime(int(date[0]), month=int(date[1]), day=int(date[2]))
+            code = session.promo_code(params)
+            session.add(code)
+            session.commit()
+            message = 'Your Code is: %s' % code.code
+        return {
+            'message': message,
+            'now': datetime.now().strftime("%Y-%m-%d"),
+            'max': c.ESCHATON.strftime('%Y-%m-%d')
+        }
+
+    @site_mappable
+    def view_promo_codes(self, session, message='', **params):
+        codes = session.query(PromoCode).all()
+        return {
+            'message': message,
+            'promo_codes': codes
+        }
+
+    def edit_code(self, session, message='', **params):
+        if 'id' and 'new_code' in params:
+            message = 'Failed for unknown reason.'
+            if params['new_code'] is not '':
+                matchingCode = session.query(PromoCode).filter(PromoCode.code == params['new_code']).first()
+                if matchingCode:
```

Comment:
I actually support you doing this check here, but this means you're both checking for matching codes when the code is edited and every time it's saved. It might make more sense to do the duplicate check only when a new code is auto-generated, and/or to split the duplicate check out into another function and call it whenever a code is edited/created by the user or by the system.